### PR TITLE
refactor(js): move generators from js/client to js/js

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

- refactor(js): move `utils/generators.ts` from `js/client` to `js/js` package
    - add necessary unit tests of `generateState` & `generateNonce` ( missing before )
    - move its unit tests as well
    - move constants about code verifier as well
    - extract `utils/scope.ts` for scope values
- chore(client):
    - add js/js to dependencies
    - fix LogtoClient unit tests
    - fix other unit tests related to the constants moved
    - fix code coverage decreased because of moving code
        - add unit tests of createRequester, discover and grantTokenByRefreshToken, 

## Linear Issue Reference

LOG-1333

## Testing

Pass all existing unit tests

---
@logto-io/eng

### CODE REVIEW

File: packages/js/src/constants.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * The length of code verifier ranges from 43 to 128.
+ * @link [Client Creates a Code Verifier](https://datatracker.ietf.org/doc/html/rfc7636#section-4.1)
+ */
+export const CODE_VERIFIER_MIN_LENGTH = 43;
+export const CODE_VERIFIER_MAX_LENGTH = 128;
+
+/**
+ * Scope - Values
+ */
+export const NAME = 'name';
```
Comment (simeng-li) [inline packages/js/src/constants.ts:22]: Do we need these two scopes?

Comment (simeng-li) [reply packages/js/src/constants.ts:22]: besides, can we use enum here?

File: packages/js/src/constants.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+export const EMAIL = 'email';
+export const OFFLINE_ACCESS = 'offline_access';
+export const OPENID = 'openid';
+
+/**
+ * Scope - Defaults
+ */
+export const DEFAULT_SCOPE_STRING = `${OPENID} ${OFFLINE_ACCESS}`;
```
Comment (simeng-li) [inline packages/js/src/constants.ts:30]: we should be able to generate the stringified scope by referencing  `DEFAULT_SCOPE_VALUES` or the other way around depends on how we gonna consume the default scopes.   But should keep a single source of truth, should not keep two different types of references of default scopes.

**Review by simeng-li — COMMENTED** [review]
there is some missing part:
1. could clean up some of the dependencies from the client package as we move out the `generators`.  i.e. js-base64 and nanoid
2. missing the test case for `generateState` and `generateNonce`  two methods.

**Comment by IceHe** [comment]
> there is some missing part:
> 
> 1. could clean up some of the dependencies from the client package as we move out the `generators`.  i.e. js-base64 and nanoid
> 
> 2. missing the test case for `generateState` and `generateNonce`  two methods.

1. Done. (see packages/client/package.json)
2. My bad. (will do it next upadte)

**Comment by simeng-li** [comment]
Better keep the test file updates and code changes within the same PR.

File: packages/js/src/constants.ts (hunk 2 above)
Comment (IceHe) [reply packages/js/src/constants.ts:30]: prefer: generate the stringified scope by referencing DEFAULT_SCOPE_VALUES

File: packages/js/src/constants.ts (hunk 1 above)
Comment (IceHe) [reply packages/js/src/constants.ts:22]: prefer enum

**Comment by IceHe** [comment]
# Problem

`codecov/project/project-client` **failed as `js/client` coverage decreased by 1.04%**.

image
( from https://app.codecov.io/gh/logto-io/js/compare/123?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr%20comments&utm_term=logto-io )

# Troubleshooting

I reviewed PR code again and compared the coverage report from both master and icehe-log-1333 branch.
It seems that **the coverage decreased just because of moving code**.

image

# Question

How to fix it? **We'd better modify codecov config?

need @simeng-li for help next Monday.

---

Reference:

- https://docs.codecov.com/docs/codecov-delta#examples
- https://docs.codecov.com/docs/unexpected-coverage-changes

**Comment by IceHe** [comment]
> there is some missing part:
> 
> 1. could clean up some of the dependencies from the client package as we move out the `generators`.  i.e. js-base64 and nanoid
> 2. missing the test case for `generateState` and `generateNonce`  two methods.

2. Done

**Review by wangsijie — APPROVED** [review]
LGTM

**Comment by IceHe** [comment]
Wait for me to fix (force-push rewrote some necessary updates).

**Comment by IceHe** [comment]
> # Problem
> `codecov/project/project-client` **failed as `js/client` coverage decreased by 1.04%**.
> 
> image ( from https://app.codecov.io/gh/logto-io/js/compare/123?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr%20comments&utm_term=logto-io )
> 
> # Troubleshooting
> I reviewed PR code again and compared the coverage report from both icehe-[log-1333](https://linear.app/silverhand/issue/LOG-1333) and master branch. It seems that **the coverage decreased just because of moving code**.
> 
> image (icehe-[log-1333](https://linear.app/silverhand/issue/LOG-1333) on the left, master on the right)
> 
> * Code coverage reports:
>   
>   * icehe-[log-1333](https://linear.app/silverhand/issue/LOG-1333): https://github.com/logto-io/js/runs/4825436420?check_suite_focus=true
>   * master: https://github.com/logto-io/js/runs/4814631919?check_suite_focus=true
> 
> # Question
> The coverage decreased just because of moving code?
> 
> * Any idea?
> 
> How to fix it?
> 
> * We'd better modify codecov config?
> * In other ways?
> 
> need @simeng-li for help next Monday.
> 
> Reference:
> 
> * https://docs.codecov.com/docs/codecov-delta#examples
> * https://docs.codecov.com/docs/unexpected-coverage-changes

Solution: Add unit tests and increase the test coverage of js/client

- test(client): discover
- test(client): grantTokenByRefreshToken
- test(client): createRequester

**Review by gao-sun — CHANGES_REQUESTED** [review]
this PR can be saved for further usage - but shall we make our road clear first?

**Comment by IceHe** [comment]
No need.

**Pull request closed without merging**